src/execution_plan/optimizations/optimize_cartesian_product.c

```diff
@@ -0,0 +1,140 @@
+	// TODO allowing AR_ExpNodes that refer directly to graph entities currently causes memory errors.
+	// This restriction should be lifted later.
+	if((f->pred.lhs->type == AR_EXP_OPERAND &&
+		f->pred.lhs->operand.type == AR_EXP_VARIADIC &&
+		f->pred.lhs->operand.variadic.entity_prop == NULL) ||
+	   (f->pred.rhs->type == AR_EXP_OPERAND &&
+		f->pred.rhs->operand.type == AR_EXP_VARIADIC &&
+		f->pred.rhs->operand.variadic.entity_prop == NULL)) return false;
+	return true;
+}
+
+// Collects all consecutive filters beneath given op.
+static OpFilter **_locate_filters(OpBase *cp) {
+	OpBase *parent = cp->parent;
+	OpFilter **filters = array_new(OpFilter *, 0);
+
+	while(parent && parent->type == OPType_FILTER) {
+		OpFilter *filter_op = (OpFilter *)parent;
+		if(_applicableFilter(filter_op->filterTree)) filters = array_append(filters, filter_op);
+		parent = parent->parent;
+	}
+
+	return filters;
+}
+
+static OpBase *_chain_cp_and_filter(const ExecutionPlan *plan, OpBase *lhs, OpBase *rhs,
+									OpBase *filter) {
+	OpBase *cp = NewCartesianProductOp(plan);
+	ExecutionPlan_AddOp(cp, lhs);
+	ExecutionPlan_AddOp(cp, rhs);
+	ExecutionPlan_AddOp(filter, cp);
+	return filter;
+}
+
+static void _optimize_cartesian_product(ExecutionPlan *plan, OpBase *cp) {
+	// Retrieve all filter operations located upstream from the Cartesian Product.
+	OpFilter **filter_ops = _locate_filters(cp);
+	uint filter_count = array_len(filter_ops);
+	/* This array will hold filters which are possible to re position in the execution plan,
+	 * but during the optimization construction they will not cause this optimization to execute. */
+	OpFilter **pending_filters = array_new(OpFilter *, filter_count);
+	OpBase *pending_filters_lookup_root = cp;
+
+	// For each stream joined by the Cartesian product, collect all entities the stream resolves.
+	int stream_count = cp->childCount;
+	rax *stream_entities[stream_count];
+	build_streams_from_op(cp, stream_entities, stream_count);
+
+	for(uint i = 0; i < filter_count; i ++) {
+		// Try to create a new dual-branched cartesian product, followed by the current filter.
+		OpFilter *filter_op = filter_ops[i];
+
+		/* A new cartesian product can be generated if both sides of the filter can be fully and
+		 * separately resolved by exactly two child streams. */
+		FT_FilterNode *f = filter_op->filterTree;
+
+		/* Make sure LHS of the filter is resolved by a stream. */
+		AR_ExpNode *lhs = f->pred.lhs;
+		uint lhs_resolving_stream = relate_exp_to_stream(lhs, stream_entities, stream_count);
+		if(lhs_resolving_stream == NOT_RESOLVED) continue;
+		/* Make sure RHS of the filter is resolved by a stream. */
+		AR_ExpNode *rhs = f->pred.rhs;
+		uint rhs_resolving_stream = relate_exp_to_stream(rhs, stream_entities, stream_count);
+		if(rhs_resolving_stream == NOT_RESOLVED) continue;
+
+		// This stream is solved by a single cartesian product child and needs to be propogate later.
+		if(lhs_resolving_stream == rhs_resolving_stream) {
```

Comment:
I don't believe that this check, and the `pending_filters` logic in general, should be necessary, as if any of the pre-optimization streams fully resolve a filter then that filter will be a child rather than the parent of the current Cartesian product. `_chain_cp_and_filter` seems to place the filter op properly without ever requiring this additional logic. The ApplyJoin optimization does use this after creating a first ValueHashJoin in queries like:
`MATCH (p1 {name: 'Ailon'}), (p2), (p3), (p4) WHERE ID(p1) = ID(p2) AND ID(p2) = ID(p3) AND ID(p3)=ID(p2) AND ID(p2)= ID(p1) AND p3.name = p4.name AND p4.name = p3.name RETURN p4.name`
But calling `applyJoin` before `optimizeCartesianProduct` allows me to place an `assert(false)` here without consequence for the test suite and a few trial queries. Are there cases I'm not considering?